### ISSUE TITLE
Severe performance degradation for tracing under 3.11

### ISSUE THREAD

**Issue by nedbat** [python/cpython#93516]
**Bug report**

Coverage.py is seeing a significant increase in overhead for tracing code in 3.11 compared to 3.10: https://github.com/nedbat/coveragepy/issues/1287

As an example:

| cov | proj | python3.10 | python3.11 | 3.11 vs 3.10 |
|----|----|----|----|----|
| none | bug1339.py | 0.184 s | 0.142 s | 76% |
| none | bm_sudoku.py | 10.789 s | 9.901 s | 91% |
| none | bm_spectral_norm.py | 14.305 s | 9.185 s | 64% |
| 6.4.1 | bug1339.py | 0.450 s | 0.854 s | 189% |
| 6.4.1 | bm_sudoku.py | 27.141 s | 55.504 s | 204% |
| 6.4.1 | bm_spectral_norm.py | 36.793 s | 67.970 s | 184% |

(This is the output of [lab/benchmark.py](https://github.com/nedbat/coveragepy/blob/master/lab/benchmark.py).)

**Your environment**

- CPython versions tested on: 3.10, 3.11 
- Operating system and architecture: MacOS, Intel

### Linked PRs
* gh-108028

**Comment by nedbat** [comment]
@pablogsal @markshannon

**Comment by Fidget-Spinner** [comment]
Related issue where cProfile slowed down 60%, and the only thing I could pinpoint was that tracing itself slowed down, not cProfile https://github.com/python/cpython/issues/93381.

**Comment by sweeneyde** [comment]
One reproducer script with only coverage.py as a dependency:

```python
from time import perf_counter
import coverage

def fib(n):
    if n <= 1:
        return n
    else:
        return fib(n-1) + fib(n-2)

t0 = perf_counter()
cov = coverage.Coverage()
cov.start()
fib(35)
cov.stop()
cov.save()
t1 = perf_counter()
print(t1 - t0)
```
</details>

Here's some output of Visual Studio performance profiler:

On the current 3.10 branch (took 13.0 seconds):

<details>

|Function Name|Total CPU \[unit, %\]|Self CPU \[unit, %\]|Module|
|-|-|-|-|
|\| - \_PyEval\_EvalFrameDefault|12606 \(99.60%\)|3594 \(28.40%\)|python310.dll|
|\| - \[External Call\] tracer.cp310-win\_amd64.pyd|2687 \(21.23%\)|1360 \(10.75%\)|tracer.cp310-win\_amd64.pyd|
|\| - maybe\_call\_line\_trace|3731 \(29.48%\)|1158 \(9.15%\)|python310.dll|
|\| - \_PyCode\_CheckLineNumber|1154 \(9.12%\)|1005 \(7.94%\)|python310.dll|
|\| - call\_trace|3322 \(26.25%\)|500 \(3.95%\)|python310.dll|
|\| - PyLong\_FromLong|358 \(2.83%\)|357 \(2.82%\)|python310.dll|
|\| - \_PyFrame\_New\_NoTrack|411 \(3.25%\)|346 \(2.73%\)|python310.dll|
|\| - \[External Code\]|12527 \(98.97%\)|324 \(2.56%\)|Multiple modules|
|\| - set\_add\_entry|291 \(2.30%\)|291 \(2.30%\)|python310.dll|
|\| - lookdict\_unicode\_nodummy|288 \(2.28%\)|288 \(2.28%\)|python310.dll|
|\| - call\_trace\_protected|2051 \(16.20%\)|282 \(2.23%\)|python310.dll|
|\| - PyDict\_GetItem|541 \(4.27%\)|262 \(2.07%\)|python310.dll|
</details>


On the current main branch (took 126.9 seconds)

<details>

|Function Name|Total CPU \[unit, %\]|Self CPU \[unit, %\]|Module|
|-|-|-|-|
|\| - \_PyEval\_EvalFrameDefault|123576 \(99.75%\)|48437 \(39.10%\)|python312.dll|
|\| - \_PyObject\_GenericGetAttrWithDict|23038 \(18.60%\)|7204 \(5.81%\)|python312.dll|
|\| - unicodekeys\_lookup\_unicode|5535 \(4.47%\)|5535 \(4.47%\)|python312.dll|
|\| - \_PyType\_Lookup|7903 \(6.38%\)|5424 \(4.38%\)|python312.dll|
|\| - PyObject\_GetAttr|26969 \(21.77%\)|3634 \(2.93%\)|python312.dll|
|\| - \_PyFrame\_Clear|3285 \(2.65%\)|2870 \(2.32%\)|python312.dll|
|\| - \_PyCode\_CheckLineNumber|6036 \(4.87%\)|2732 \(2.21%\)|python312.dll|
</details>

The first difference I notice is that there's a lot of overhead in some call to \_PyObject\_GenericGetAttrWithDict.

**Comment by sweeneyde** [comment]
On main, if I add `printf("lookup %s\n", (const char *)PyUnicode_DATA(key));` to the top of `unicodekeys_lookup_unicode` and run that `fib` script, I get the following hot loop repeated over and over:

<details>

```
lookup fib
lookup THIS_FILE
lookup stopped
lookup should_start_context
lookup started_context
lookup _activity
lookup data_stack
lookup cur_file_data
lookup cur_file_name
lookup last_line
lookup cur_file_name
lookup cur_file_data
lookup RESUME
lookup last_line
lookup _cached_bound_method_trace
lookup THIS_FILE
lookup stopped
lookup cur_file_data
lookup trace_arcs
lookup cur_file_data
lookup last_line
lookup _cached_bound_method_trace
lookup THIS_FILE
lookup stopped
lookup cur_file_data
lookup trace_arcs
lookup cur_file_data
lookup last_line
lookup _cached_bound_method_trace
```
</details>

**Comment by sweeneyde** [comment]
I now realize this is probably due to using the Python tracer rather than the C tracer, will try again.

**Comment by sweeneyde** [comment]
Okay, now with `Code coverage for Python, version 6.4.1 with C extension.`

<details>
<summary>Script used:</summary>

```python
import coverage
from time import perf_counter

def fib(n):
    if n <= 1:
        return n
    else:
        return fib(n-1) + fib(n-2)

t0 = perf_counter()
cov = coverage.Coverage()
cov.start()
fib(35)
cov.stop()
cov.save()
t1 = perf_counter()
print(t1 - t0)
```
</details>



<details>
<summary>Profile on 3.10 branch: took 13.3 seconds</summary>

Functions taking more than 1% of CPU:

|Function Name|Total CPU \[unit, %\]|Self CPU \[unit, %\]|Module|
|-|-|-|-|
|\| - \_PyEval\_EvalFrameDefault|12372 \(99.78%\)|3540 \(28.55%\)|python310.dll|
|\| - \[External Call\] tracer.cp310-win\_amd64.pyd|2769 \(22.33%\)|1474 \(11.89%\)|tracer.cp310-win\_amd64.pyd|
|\| - maybe\_call\_line\_trace|3872 \(31.23%\)|1206 \(9.73%\)|python310.dll|
|\| - \_PyCode\_CheckLineNumber|1168 \(9.42%\)|988 \(7.97%\)|python310.dll|
|\| - call\_trace|3431 \(27.67%\)|548 \(4.42%\)|python310.dll|
|\| - PyLong\_FromLong|370 \(2.98%\)|369 \(2.98%\)|python310.dll|
|\| - \_PyFrame\_New\_NoTrack|385 \(3.11%\)|338 \(2.73%\)|python310.dll|
|\| - lookdict\_unicode\_nodummy|324 \(2.61%\)|324 \(2.61%\)|python310.dll|
|\| - call\_trace\_protected|2107 \(16.99%\)|287 \(2.31%\)|python310.dll|
|\| - frame\_dealloc|270 \(2.18%\)|270 \(2.18%\)|python310.dll|
|\| - set\_add\_entry|264 \(2.13%\)|264 \(2.13%\)|python310.dll|
|\| - PyDict\_GetItem|579 \(4.67%\)|255 \(2.06%\)|python310.dll|
|\| - \_PyEval\_MakeFrameVector|633 \(5.11%\)|248 \(2.00%\)|python310.dll|
|\| - \_PyEval\_Vector|12372 \(99.78%\)|196 \(1.58%\)|python310.dll|
|\| - PyLineTable\_NextAddressRange|180 \(1.45%\)|180 \(1.45%\)|python310.dll|
|\| - call\_function|12372 \(99.78%\)|169 \(1.36%\)|python310.dll|
|\| - PyObject\_RichCompare|457 \(3.69%\)|169 \(1.36%\)|python310.dll|
|\| - binary\_op1|486 \(3.92%\)|157 \(1.27%\)|python310.dll|
|\| - long\_richcompare|207 \(1.67%\)|157 \(1.27%\)|python310.dll|
</details>

<details>
<summary>Profile on main branch: took 23.0 seconds</summary>

Functions taking more than 1% of CPU:

|Function Name|Total CPU \[unit, %\]|Self CPU \[unit, %\]|Module|
|-|-|-|-|
|\| - \_PyEval\_EvalFrameDefault|22180 \(99.87%\)|4448 \(20.03%\)|python312.dll|
|\| - \_PyCode\_CheckLineNumber|5389 \(24.27%\)|2583 \(11.63%\)|python312.dll|
|\| - maybe\_call\_line\_trace|6461 \(29.09%\)|1966 \(8.85%\)|python312.dll|
|\| - retreat|2242 \(10.10%\)|1644 \(7.40%\)|python312.dll|
|\| - \[External Call\] tracer.cp312-win\_amd64.pyd|6274 \(28.25%\)|1428 \(6.43%\)|tracer.cp312-win\_amd64.pyd|
|\| - get\_line\_delta|1162 \(5.23%\)|1162 \(5.23%\)|python312.dll|
|\| - unicodekeys\_lookup\_unicode|844 \(3.80%\)|728 \(3.28%\)|python312.dll|
|\| - \_Py\_dict\_lookup|1476 \(6.65%\)|632 \(2.85%\)|python312.dll|
|\| - call\_trace|9802 \(44.14%\)|630 \(2.84%\)|python312.dll|
|\| - siphash13|442 \(1.99%\)|442 \(1.99%\)|python312.dll|
|\| - PyUnicode\_New|745 \(3.35%\)|376 \(1.69%\)|python312.dll|
|\| - pymalloc\_alloc|362 \(1.63%\)|362 \(1.63%\)|python312.dll|
|\| - \_PyType\_Lookup|1786 \(8.04%\)|319 \(1.44%\)|python312.dll|
|\| - set\_add\_entry|271 \(1.22%\)|271 \(1.22%\)|python312.dll|
|\| - PyDict\_GetItem|824 \(3.71%\)|242 \(1.09%\)|python312.dll|
|\| - call\_trace\_protected|8562 \(38.55%\)|236 \(1.06%\)|python312.dll|
|\| - PyNumber\_Subtract|449 \(2.02%\)|221 \(1.00%\)|python312.dll|
</details>

**Comment by sweeneyde** [comment]
This seems to suggest to me that this could be made much faster via a dedicated C API function:

https://github.com/nedbat/coveragepy/blob/master/coverage/ctracer/util.h#L41
```c
#define MyCode_GetCode(co)      (PyObject_GetAttrString((PyObject *)(co), "co_code"))
```

**Comment by sweeneyde** [comment]
Indeed, on main, if I add `printf("lookup %s\n", (const char *)PyUnicode_DATA(key));` to the top of `unicodekeys_lookup_unicode` and run that fib script, I get the following hot loop repeated over and over:

```
lookup fib
lookup C:\Users\sween\Source\Repos\cpython2\cpython\cover.py
lookup C:\Users\sween\Source\Repos\cpython2\cpython\cover.py
lookup co_code
lookup fib
lookup C:\Users\sween\Source\Repos\cpython2\cpython\cover.py
lookup C:\Users\sween\Source\Repos\cpython2\cpython\cover.py
lookup co_code
...
```

Each of these calls to `PyObject_GetAttrString(..., "co_code")` requires `PyUnicode_FromString->unicode_decode_utf8->PyUnicode_New->_PyObject_Malloc`, followed by `PyObject_GetAttr->_PyObject_GenericGetAttrWithDict->_PyType_Lookup->find_name_in_mro->(unicode_hash+_Py_Dict_Lookup)->unicodekeys_lookup_unicode`, where it should be just a C API function call.

**Comment by sweeneyde** [comment]
Ah! `PyCode_GetCode` exists since https://github.com/python/cpython/pull/92168

**Comment by markshannon** [comment]
Some performance degradation under tracing is expected, but not as much as reported.
This is a deliberate tradeoff: faster execution when not tracing, for slower execution when tracing.

From the profiles that @sweeneyde gathered it looks like we are seeing slowdowns in:
* _PyEval_EvalFrameDefault, which is expected
* Calculation of line numbers. This is probably due to the new line table format.

The new line number table gives better error messages without taking up too much extra space, but it is more complex and thus slower to parse.
We should be able to speed it up, but don't expect things to be as fast as 3.10. Sorry.

Once https://github.com/nedbat/coveragepy/pull/1394 is merged, we can look at the profile again to see if much has changed.

**Comment by sweeneyde** [comment]
<details>
<summary>New profile of fib_cover.py on main branch, with the PyCode_GetCode addition</summary>

|Function Name|Total CPU \[unit, %\]|Self CPU \[unit, %\]|Module|
|-|-|-|-|
|\| - \_PyEval\_EvalFrameDefault|19239 \(99.66%\)|4481 \(23.21%\)|python312.dll|
|\| - \_PyCode\_CheckLineNumber|5344 \(27.68%\)|2676 \(13.86%\)|python312.dll|
|\| - maybe\_call\_line\_trace|6397 \(33.14%\)|1997 \(10.34%\)|python312.dll|
|\| - retreat|2197 \(11.38%\)|1629 \(8.44%\)|python312.dll|
|\| - \[External Call\] tracer.cp312-win\_amd64.pyd|3073 \(15.92%\)|1360 \(7.04%\)|tracer.cp312-win\_amd64.pyd|
|\| - get\_line\_delta|1038 \(5.38%\)|1038 \(5.38%\)|python312.dll|
|\| - call\_trace|6630 \(34.34%\)|605 \(3.13%\)|python312.dll|
|\| - unicodekeys\_lookup\_unicode|549 \(2.84%\)|548 \(2.84%\)|python312.dll|
|\| - \_Py\_dict\_lookup|1060 \(5.49%\)|511 \(2.65%\)|python312.dll|
|\| - PyDict\_GetItem|929 \(4.81%\)|266 \(1.38%\)|python312.dll|
|\| - set\_add\_entry|265 \(1.37%\)|265 \(1.37%\)|python312.dll|
|\| - PyNumber\_Subtract|447 \(2.32%\)|226 \(1.17%\)|python312.dll|
|\| - call\_trace\_protected|5408 \(28.01%\)|222 \(1.15%\)|python312.dll|
|\| - initialize\_locals|199 \(1.03%\)|199 \(1.03%\)|python312.dll|
</details>

I wonder: would there be any way to retain some specializations during tracing? Some specialized opcodes are certainly incorrect to use during tracing, e.g., `STORE_FAST__LOAD_FAST`. However, couldn't others be retained, e.g. `LOAD_GLOBAL_MODULE`, or any specialized opcode that covers only one unspecialized opcode?

**Comment by pablogsal** [comment]
> We should be able to speed it up, but don't expect things to be as fast as 3.10. Sorry.

Well, two times slower than 3.10, as some of @nedbat's benchmarks show, is not acceptable in my opinion. We should try to get this below 20% at the very least.

**Comment by markshannon** [comment]
The problem with saying that something is unacceptable, is that implies there is an acceptable alternative.
What is that alternative?

What does it mean making this a release blocker? Are you not going to release 3.11 if coverage is >20% slower?

I don't think cutoffs, such as 20%, are constructive. We should do as well as we are reasonably able, regardless of whether we achieve some arbitrary number.
I expect that will be better than a 20% slowdown, but I'm not making any pronouncements.

**Comment by pablogsal** [comment]
> What is that alternative?

I don't know, but there is always the possibility to revert the changes that made coverage slower (worst case scenario).

>  Are you not going to release 3.11 if coverage is >20% slower?

I will check with the SC, but I can already tell you that I won't release 3.11 if is 2 times slower in coverage, unless the SC mandates otherwise.

> I don't think cutoffs, such as 20%, are constructive. We should do as well as we are reasonably able, regardless of whether we achieve some arbitrary number.

The problem here is that this tradeoff was not discussed anywhere or advertised and no one had the possibility to pronounce or to make tradeoffs or otherwise, and therefore we cannot just "decide" out of the blue what is and what is not acceptable because the community didn't had the change to object.

**Comment by markshannon** [comment]
The largest contributor to the slowdown is computing line numbers. The root cause of that is PEP 657. The proximate cause is my PR to reduce the memory consumed of PEP 657.
Would you revert PEP 657, or just accept a large increase in memory use?

I think it would be a lot more helpful if we fixed the issue, rather than making dire pronouncements.

> the community didn't had the change to object.

Everything we do is public. I even opened  https://github.com/nedbat/coveragepy/issues/1287 back in November, so that @nedbat was aware that there might be slowdowns to coverage.py

**Comment by pablogsal** [comment]
> The largest contributor to the slowdown is computing line numbers. The root cause of that is PEP 657.

That's not true. When line numbers and column offsets where separated, PEP 657 had 0 overhead on computing line numbers

**Comment by pablogsal** [comment]
In any case, there is no point to discuss in these terms because I am not the one making decisions here. As RM my only authority is to consider the **current** slowdown a blocker and ask for a bigger discussion to be taken if we cannot fix the issue.

Let's see what we can do and then let's discuss how to proceed once we understand our options.

**Comment by markshannon** [comment]
> The root cause was merging line numbers and column offsets so you need to pay for decoding both just to get line numbers.

No, that was the proximate cause. It wouldn't have been necessary if not for PEP 657.
I didn't implement it to make things slower, I implemented it because column and endline tables were consuming a lot of memory.

Claiming that PEP 657 had zero overhead is not true. The original implementation had a large overhead in terms on memory use.

**Comment by pablogsal** [comment]
> Claiming that PEP 657 had zero overhead is not true. The original implementation had a large overhead in terms on memory use.

The claim was that PEP 657 had zero overhead on time, not memory.

**Comment by gpshead** [comment]
Marking this as a release or deferred blocker is fine and entirely what I'd expect any RM to want to do. That just means we collect more information and make a decision before we exit beta stages.

I don't _personally_ have a problem with coverage being slower, even by 2x. It's more about understanding what all is impacted by this and how so we can explain our measured rationale to users.  ie: Look beyond just `coverage` which should only be a Developer/CI time tool rather than something slowing people's "production" release tool/application/notebook code.

Q1: What else uses tracing?

Q1b: What of those use it at other than development time?

These questions are informational and do not block performance work on reducing the impact to tracing.

**Comment by sweeneyde** [comment]
How bad would it be to store an extra field into code objects that is the "decompressed" line number table, lazily allocated for only those code objects which are actually traced? Could it occupy `co_extra` or similar?

**Comment by neonene** [comment]
Does this issue focus on C tracer cases?

**Comment by Fidget-Spinner** [comment]
> How bad would it be to store an extra field into code objects that is the "decompressed" line number table, lazily allocated for only those code objects which are actually traced? Could it occupy `co_extra` or similar?

In https://github.com/python/cpython/pull/93383#issuecomment-1145729820 I benchmarked adding an additional field to the code object and there was no slowdown in pyperformance. I do not know how memory consumption is affected.

I'm not sure co_extra is the right place for that info considering its for consumers of the PEP 523 API. Maybe I can merge your idea into a simple lazily created `_PyCodeTracingInfo *` C struct?

```C
PyCodeObject {
    ...
     void *co_tracing_info
}

struct _PyCodeTracingInfo {
     // cached code bytes object
    PyObject *co_code;
     // decompressed line nunber table
}
```

**Comment by markshannon** [comment]
> How bad would it be to store an extra field into code objects that is the "decompressed" line number table, lazily allocated for only those code objects which are actually traced? Could it occupy `co_extra` or similar?

An early design for the compressed line number table did just this.
My concern, then and now, is that the uncompressed table is a *lot* bigger, and never goes away.

It isn't just tracing that needs line numbers, tracebacks need them too. 
Which means that the uncompressed table would be created every time a code object ends up in a traceback.

I think we should see how much faster we can make tracing with the compressed table first. If that isn't good enough, then we can reconsidered the decompressed table.

**Comment by markshannon** [comment]
@gpshead 

> Q1: What else uses tracing?

Debuggers and profiling tools, like cProfile.

Q1b: What of those use it at other than development time?

Hopefully, no one. Tracing has always imposed a considerable overhead.
Anyone using a profiler in production should be using a statistical profiler.

**Comment by pablogsal** [comment]
> I think we should see how much faster we can make tracing with the compressed table first. If that isn't good enough, then we can reconsidered the decompressed table.

Maybe we could leave line numbers uncompressed and compress the rest?

Alternatively, we can leave uncompressed just the line numbers *if and only if* we detect a tracing or profiling function? This way the extra memory is only paid if someone is profiling or tracing.

**Comment by markshannon** [comment]
> Maybe we could leave line numbers uncompressed and compress the rest?

That's going to be complicated. Too complicated for a beta release, IMO.

> Alternatively, we can leave uncompressed just the line numbers if and only if we detect a tracing or profiling function? This way the extra memory is only paid if someone is profiling or tracing.

If we add a small/large flag, then the extra memory will be 2 bytes for instructions for files < 64k lines (basically all code), and 4 bytes per instruction for giant files.
I'll try it out.

**Comment by markshannon** [comment]
I've been experimenting with streamlining the `advance` and `maybe_call_line_trace` functions. I'm getting roughly a 10% speedup, but a flat table lookup will be way faster.

**Comment by markshannon** [comment]
Actually if we store the line delta from `co_firstline`, many functions will only need 1 bytes per code unit, which should be fine.

**Comment by markshannon** [comment]
Looks like the bug is in the compiler front end. It is producing a instruction where `lineno == end_lineno` and `s.end_col_offset < col_offset`.

**Comment by nedbat** [comment]
I'm very interested in helping to test people's proposed fixes, but am losing track of what is being changed where.  We have #93493, which I have tried.  Are there other pull requests in flight to try?

**Comment by pablogsal** [comment]
I would like to see how much we gain back by reverting 944fffee8916cb94321fa33cd3a43f4108717746 (not that we are planning to do it, just want to know how much the problem is due to that).

I can prepare a PR for you to test later today but in the interim you can try reverting yourself (it doesn't cleanly revert, but the conflicts are due to automatically generated files mainly that you can regenerate by running `make regen-all`).

**Comment by Fidget-Spinner** [comment]
> I would like to see how much we gain back by reverting 944fffee8916cb94321fa33cd3a43f4108717746 (not that we are planning to do it, just want to know how much the problem is due to that).

IIUC, Kumar benchmarked that commit here and got a 1.3x slowdown.

https://github.com/python/pyperformance/pull/213#issuecomment-1147699232

**Comment by pablogsal** [comment]
> IIUC, Kumar benchmarked that commit here and got a 1.3x slowdown before and after.

That's in pyperformance, no? I meant to see how much it performs in coverage

**Comment by markshannon** [comment]
Looking at the output of perf (using an optimized, but no pgo, lto build) and the code, I see the following reasons for the slowdown:

1. Increased overhead in the `_PyEval_EvalFrameDefault` determining whether the event is "call", possible "line" or no event.
2. Increased overhead in the `_PyEval_EvalFrameDefault` as we no longer specialize when tracing.
3. Increased overhead in `maybe_call_line_trace` finding the first traceable instruction
4. Large slowdown in determining the line number from the index.

All but 2 can be fixed without any negative impact on performance when not tracing.
Fixing 4 will increase memory use when tracing, but when not tracing, will only need a tiny bit of extra memory.

Specializing during tracing, increases the overhead of dispatching when not-tracing resulting in a ~1% slowdown.

**Comment by kumaraditya303** [comment]
FYI, I have added a coverage [^1] benchmark to [pyperformance](https://github.com/python/pyperformance) to ease benchmarking coverage / tracing performance.

[^1]:https://github.com/python/pyperformance/pull/213

**Comment by markshannon** [comment]
https://github.com/python/cpython/pull/93769

**Comment by Fidget-Spinner** [comment]
@pablogsal can I trouble you to regen the ABI for #93493 please? It should be the last PR for this issue and it's a roughly 10% improvement in coverage.py's benchmark.

**Comment by Fidget-Spinner** [comment]
Ooh thanks for writing that new section in the devguide! I'll try and report back.

**Comment by pablogsal** [comment]
I plan to try to release beta 4 tomorrow, so it would be great if we can get a decision with https://github.com/python/cpython/pull/93493

**Comment by pablogsal** [comment]
> I just stumbled upon this, so, I ran my profilings from `pydevd` -- in the pure-python tracer (I'm still working on the cython and frame eval modes), the numbers I get is that 3.11 is 50-60% slower than 3.10 in general.
> 
> -- I must say I wasn't expecting a speedup when tracing is enabled in 3.11, but I also wasn't expecting such a big slowdown... 
> 
> The raw numbers for the benchmarks I have are below:
> 
> ```
> Python 3.11 beta 3
> method_calls_with_breakpoint: 0.383s (/ 0.250 = 1.53: slower)
> method_calls_without_breakpoint: 0.381s (/ 0.247 = 1.54: slower)
> method_calls_with_step_over: 0.378s (/ 0.236 = 1.60: slower)
> method_calls_with_exception_breakpoint: 0.375s (/ 0.249 = 1.50: slower)
> global_scope_1_with_breakpoint: 0.854s (/ 0.557 = 1.53: slower)
> global_scope_2_with_breakpoint: 0.364s (/ 0.238 = 1.52: slower)
> 
> Python 3.11 (repository: 41e4b426ce)
> method_calls_with_breakpoint: 0.380s 
> method_calls_without_breakpoint: 0.368s 
> method_calls_with_step_over: 0.357s 
> method_calls_with_exception_breakpoint: 0.380s 
> global_scope_1_with_breakpoint: 0.920s 
> global_scope_2_with_breakpoint: 0.337s 
> 
> Python 3.10.0
> method_calls_with_breakpoint: 0.250s 
> method_calls_without_breakpoint: 0.247s 
> method_calls_with_step_over: 0.236s 
> method_calls_with_exception_breakpoint: 0.249s 
> global_scope_1_with_breakpoint: 0.557s 
> global_scope_2_with_breakpoint: 0.238s 
> ```

@fabioz are you using the latest released beta or the tip of the 3.11 branch?

**Comment by fabioz** [comment]
> @fabioz are you using the latest released beta or the tip of the 3.11 branch?

@pablogsal I've tried with both (the benchmarks I posted are for Python 3.11 beta 3 and Python 3.11 tip which I compiled locally -- note that I did not compile with PGO just release -- i.e.: `PCbuild\build.bat -e -p x64`)... The differences from the tip to the release version weren't significant (some were a bit slower, some were a bit faster some almost the same, so, I'd say it's in the noise range).

**Comment by hodgestar** [comment]
As a random Python developer, I would be unhappy with a 20% slowdown in the speed of coverage, under the assumption that that would make all of my continuous integration runs take 20% longer. :/

**Comment by nedbat** [comment]
I did a benchmark, and the latest tip of 3.11 (commit c966e0816e) is looking better, though still slower than 3.10:
| cov                  | proj                 | v3.10.5              | v3.11.0b3            | 3.11_tip             | 3.11b3 vs 3.10       | 3.11 tip vs 3.10     |
| :---                 | :---                 | ---:                 | ---:                 | ---:                 | ---:                 | ---:                 |
| 6.4.1                | bug1339.py           | 0.399 s              | 0.832 s              | 0.469 s              | 208.41%              | 117.42%              |
| 6.4.1                | bm_sudoku.py         | 26.273 s             | 55.805 s             | 34.877 s             | 212.40%              | 132.75%              |
| 6.4.1                | bm_spectral_norm.py  | 32.660 s             | 71.043 s             | 44.546 s             | 217.52%              | 136.39%              |

**Comment by gpshead** [comment]
Quite frankly, those 3.11 tip numbers look well within reason to accept.

Tracing is a relatively rare operation in terms of all CPU cycles consumed by CPython processes on all computers around the world. It is worth sacrificing some tracing performance there if deemed necessary in order to gain performance where it is actually statistically significant: in deployed python programs.

**Comment by alex** [comment]
As a top line conclusion, I agree that with these improvements this no longer needs to block the release.

That said, I think it's important to note that coverage performance has a substantial impact on developer experience. For many projects tests-with-coverage is a blocking element of their CI pipeline, so this performance ultimately dictates their iteration cycle time. It cannot happen for 3.11, but I think on a longer timeline it's worth exploring what can be done to improve coverage's performance (which perhaps is another external tool like https://github.com/plasma-umass/slipcover) -- particularly if it's going to be in tension with other optimizations we want to land for the general case!

**Comment by Fidget-Spinner** [comment]
We still have Mark's PR #94231. Once that's merged I suspect the benchmark numbers for 3.11 tip will drop to below a 30% slowdown over 3.10.

**Comment by pablogsal** [comment]
@nedbat @fabioz Can you repeat the benchmarks with #94231?

**Comment by nedbat** [comment]
Yup, I just did that. About the same as without the change:
| cov                  | proj                 | v3.10.5              | v3.11.0b3            | 94231                | 3.11b3 vs 3.10       | 94231 vs 3.10        |
| :---                 | :---                 | ---:                 | ---:                 | ---:                 | ---:                 | ---:                 |
| 6.4.1                | bug1339.py           | 0.517 s              | 1.155 s              | 0.572 s              | 223.22%              | 110.64%              |
| 6.4.1                | bm_sudoku.py         | 29.552 s             | 59.221 s             | 36.008 s             | 200.40%              | 121.85%              |
| 6.4.1                | bm_spectral_norm.py  | 33.720 s             | 77.685 s             | 49.937 s             | 230.38%              | 148.09%              |

**Comment by Fidget-Spinner** [comment]
I now agree with Gregory and Alex that the performance regression is now within acceptable levels. IMO we can leave this issue open but it shouldn't block the 3.11 release.

**Comment by fabioz** [comment]
I just reran the performance for `pydevd` with the current tip. The numbers haven't changed from the last run (it's still in the range of 45 -> 70 percent slower depending on the use case, so, at least for `pydevd` the current tip measurements are very close to the beta 3 measurements).

Benchmark | Python 3.10 | Python 3.11 tip ( abf5f5c5d7 ) | Slower than 3.10 in %
-- | -- | -- | --
method_calls_with_breakpoint | 0,25 | 0,363 | 45,20%
method_calls_without_breakpoint | 0,247 | 0,375 | 51,82%
method_calls_with_step_over | 0,236 | 0,376 | 59,32%
method_calls_with_exception_breakpoint | 0,249 | 0,389 | 56,22%
global_scope_1_with_breakpoint | 0,557 | 0,937 | 68,22%
global_scope_2_with_breakpoint | 0,238 | 0,343 | 44,12%

**Comment by jpe** [comment]
Hmm, I may be confusing myself but I'm seeing tip take about 50% of the time of b3 when running one of our test files. I'm testing with a C trace function that doesn't do anything except return 0. Looking at the diffs since b3, I think performance partially depends on the number of code objects.

**Comment by pablogsal** [comment]
Thanks for the report @P403n1x87! Unfortunately, the results you show are not very useful (is mostly what we are already getting by running `perf`) because what we need to understand is what linear or non-linear combination of the changes we did are affecting the coverage numbers. The flame graphs and tables show the difference between 3.10.5 and 3.11 which are two completely code paths for everything (they don't even have the same instructions). What we need to understand is the relative cost of the coverage-relative machinery before and after every change that we did that could affect this. Unfortunately tracing over all the interpreter is too unstable and noisy to give us insight.

**Comment by StanFromIreland** [comment]
This needs to be removed from the differed/release blocker project.

**Comment by markshannon** [comment]
There are going to be no direct fixes for this issue, but I expect it to become a non-issue over the course of the 3.15 development cycle for two reasons:
* Now that sys.monitoring supports efficient branch monitoring, coverage.py will use it by default and the performance impact on coverage will go. And really, it is only coverage that this matters for. All debuggers now use sys.monitoring.
* As we move to tagged integers the boxing/unboxing overhead for line events will disappear, which is the main cause of the slowdown.

@nedbat is my statement about coverage.py correct?
